src/js/libgeo/GeoBasics.js

```diff
@@ -360,75 +360,91 @@ function addElementNoProof(el) {
+        return;
+    }
 
-    for (i = 0; i < csgeo.gslp.length; i++) {
-        el = csgeo.gslp[i];
+    // build dependency tree
+    var depTree = {};
+    var cskeys = Object.keys(csgeo.csnames);
+    cskeys.forEach(function(name) {
+        var el = csgeo.csnames[name];
+        if (!el.hasOwnProperty("args")) return;
+        var args = el.args;
 
-        if (el.name === name) {
-            if (debug) console.log("Removed element from gslp " + name);
-            csgeo.gslp.splice(i, 1);
-        }
-    }
+        args.forEach(function(argn) {
+            if (!depTree.hasOwnProperty(argn)) {
+                depTree[argn] = {};
+            }
+            depTree[argn][name] = true;
+        });
+    });
 
-    for (i = 0; i < csgeo.free.length; i++) {
-        el = csgeo.free[i];
+    // recursively find all dependencies
+    var recFind = function(elname, map) {
+        map[elname] = true;
+        if (!depTree.hasOwnProperty(elname)) return map;
 
-        if (el.name === name) {
-            if (debug) console.log("Removed element from free " + name);
-            csgeo.free.splice(i, 1);
+        for (var dn in depTree[elname]) {
+            if (!map[dn]) {
+                recFind(dn, map);
+            }
         }
-    }
 
-    for (i = 0; i < csgeo.points.length; i++) {
-        el = csgeo.points[i];
+        return map;
+    };
 
-        if (el.name === name) {
-            if (debug) console.log("Removed element from points " + name);
-            csgeo.points.splice(i, 1);
-        }
-    }
+    // collect all objects to delete
+    var delArr = recFind(name, {});
 
-    for (i = 0; i < csgeo.lines.length; i++) {
-        el = csgeo.lines[i];
+    removeAllElements(delArr);
 
-        if (el.name === name) {
-            if (debug) console.log("Removed element from lines " + name);
-            csgeo.lines.splice(i, 1);
-        }
-    }
+}
 
-    for (i = 0; i < csgeo.conics.length; i++) {
-        el = csgeo.conics[i];
+function removeAllElements(nameMap) {
+    var i, el, debug = false;
+    var keys = Object.keys(nameMap);
+    if (debug) console.log("Remove elements: " + String(keys));
 
-        if (el.name === name) {
-            if (debug) console.log("Removed element from conics " + name);
-            csgeo.conics.splice(i, 1);
-        }
-    }
 
-    // remove from sets
-    //
-    // define function with closure which compares the name
-    var nameCmp = function(cmpName) {
+    var nameCmp = function(cmpMap, arrn) {
         return function(setel) {
-            if (setel.name === cmpName) {
-                if (debug) console.log("Removed element " + name + " from set of " + sname);
+            if (cmpMap[setel.name]) {
+                if (debug) console.log("Removed element " + setel.name + " from " + arrn);
                 return false;
             } else return true;
         };
     };
 
+    // update incidences
+    keys.forEach(function(name) {
```

Comment:
Honestly, it is not clear to me, why this might not cause a quadratic running time in pathological cases.

But heuristically, I am willing to believe that this runs smooth enough in most cases because there are only a few incidences.

Reply:
The best way to avoid quadratic runtime is to mark all elements that are to be deleted and not recurse for the marked ones. 

Again, sorry for just dropping comments and not implementing it properly, but that's the only thing I can do within the 2 minutes of attention span…